File: src/Historical/Pascal/factory_method.pas
program FactoryMethod;

{$mode objfpc}{$H+}

type
  TDatabaseAction = procedure;

  TDatabase = record
    Connect: TDatabaseAction;
    Query: TDatabaseAction;
  end;

  TFactoryMethod = function: TDatabase;

procedure PostgresConnect;
begin
  WriteLn('PostgreSQL connect');
end;

procedure PostgresQuery;
begin
  WriteLn('PostgreSQL query');
end;

procedure MySqlConnect;
begin
  WriteLn('MySQL connect');
end;

procedure MySqlQuery;
begin
  WriteLn('MySQL query');
end;

function CreatePostgres: TDatabase;
begin
  Result.Connect := @PostgresConnect;
  Result.Query := @PostgresQuery;
end;

function CreateMySql: TDatabase;
begin
  Result.Connect := @MySqlConnect;
  Result.Query := @MySqlQuery;
end;

procedure UseDatabase(CreateDatabase: TFactoryMethod);
var
  Database: TDatabase;
begin
  Database := CreateDatabase();
  Database.Connect();
  Database.Query();
end;

begin
  UseDatabase(@CreatePostgres);
  UseDatabase(@CreateMySql);
end.

File: src/Historical/Fortran/factory_method.f90
program factory_method
  implicit none

  abstract interface
    integer function factory_method_proc()
    end function factory_method_proc
  end interface

  call use_database(create_postgres)
  call use_database(create_mysql)

contains

  integer function create_postgres()
    create_postgres = 1
  end function create_postgres

  integer function create_mysql()
    create_mysql = 2
  end function create_mysql

  subroutine use_database(create_database)
    procedure(factory_method_proc) :: create_database
    integer :: database

    database = create_database()
    select case (database)
    case (1)
      print '(A)', 'PostgreSQL connect'
      print '(A)', 'PostgreSQL query'
    case (2)
      print '(A)', 'MySQL connect'
      print '(A)', 'MySQL query'
    case default
      error stop 'unknown database'
    end select
  end subroutine use_database

end program factory_method
